### PR TITLE
added popper sampler and pooler

### PR DESCRIPTION
**Pull request by musslick**
## Description

This feature implements a pooler and sampler for the Popper Experimentalist. 

The pooler generates samples for independent variables with the objective of maximizing the (approximated) loss of the model. The samples are generated by first training a neural network to approximate the loss of a model for all patterns in the training data. Once trained, the network is then inverted to generate samples that maximize the approximated loss of the model.

The sampler leverages the pooler to select samples that are closest do "ideal" candidates.

The PR comes with two unit tests, one for a classification model and one for a regression model. 

### Type of change:
- New feature (non-breaking change which adds functionality)

### Features: 
- Popper Net Pooler (see description above)
- Popper Net Sampler (see description above)
- unit test for classification model
- unit test for the regression model

### Questions: 
- In the doc string of the sampler, I wanted to cross-reference the popper pooler function but wasn't sure how to do that.

### [Optional] Screenshots:
There are some helper functions (in verbose mode) that illustrate the quality of the Popper Network, i.e., how well it can approximate the error of the model. Here's an example:

Training curve of the popper network:
image

The predicted error of the model versus the actual (target) error of the model:

image

### CODE REVIEW

**Review by hollandjg — COMMENTED** [review]
This is great! I've worked through this a bit and I think there are a couple of places we could streamline it.
- I think the `popper_sampler` could be a more generic "nearest value to a list of values we want to allow" sampler, which might be valuable in other contexts. We can just drop it into a pipeline and use it there. I'll show you that tomorrow.
- I'm not sure the sampler is doing the right thing yet – I think you could craft inputs which give the wrong outputs. I'll have a go at that before we meet and show you.
- It's not clear to me from the code what the poppernet_pooler is doing, and I think here again we might be better separating the categorical from the regression pooler, as they're in some ways fundamentally different. We can discuss that tomorrow.

File: autora/experimentalist/sampler/poppernet.py [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+
+        plt.xlabel("x")
+        plt.ylabel("y")
+        plt.legend()
+        plt.show()
+
+        plt.plot(losses)
+        plt.xlabel("epoch")
+        plt.ylabel("loss")
+        plt.show()
```
[the remark this answers is not in the document]
Comment (musslick) [reply autora/experimentalist/sampler/poppernet.py:155]: Hey John,

Thanks so much for taking a look. All of your suggestions sound (as always) great. 

I have only one push-back on whether to write the pooler for both categorial and regression samplers. I would argue that the problem of identifying useful IVs via the popper approach is identical across the two regression and classification problems: it's only about predicting the model error. There is also a large benefit from a user's perspective of not having to care which pooler to use depending on whether the model is doing classification vs. regression. Let's discuss it tomorrow!

**Review by hollandjg — APPROVED** [review]
🎅🎄
